{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6ccd9611",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import os\n",
    "import pymongo\n",
    "from bson.objectid import ObjectId\n",
    "from bson.dbref import DBRef\n",
    "from Bio import Entrez\n",
    "from Bio import GenBank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "602fc3e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Connecting to mongodb and getting db\n",
    "client = pymongo.MongoClient(\"localhost\", 27017, username=\"admin\", password=\"admin\")\n",
    "db = client['taller_final']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8e0ef4e",
   "metadata": {},
   "source": [
    "# Transposable elements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0b67c20b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>te_id</th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>sequence_ontology</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>classification</th>\n",
       "      <th>sequence_ontology_v2</th>\n",
       "      <th>identity</th>\n",
       "      <th>method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ptg000002l_0</td>\n",
       "      <td>ptg000002l</td>\n",
       "      <td>EDTA</td>\n",
       "      <td>Gypsy_LTR_retrotransposon</td>\n",
       "      <td>15839</td>\n",
       "      <td>17218</td>\n",
       "      <td>4477</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>LTR/Gypsy</td>\n",
       "      <td>SO:0002265</td>\n",
       "      <td>0.753</td>\n",
       "      <td>homology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ptg000002l_1</td>\n",
       "      <td>ptg000002l</td>\n",
       "      <td>EDTA</td>\n",
       "      <td>Gypsy_LTR_retrotransposon</td>\n",
       "      <td>17520</td>\n",
       "      <td>19985</td>\n",
       "      <td>12445</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>LTR/Gypsy</td>\n",
       "      <td>SO:0002265</td>\n",
       "      <td>0.835</td>\n",
       "      <td>homology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ptg000002l_2</td>\n",
       "      <td>ptg000002l</td>\n",
       "      <td>EDTA</td>\n",
       "      <td>Gypsy_LTR_retrotransposon</td>\n",
       "      <td>19987</td>\n",
       "      <td>20451</td>\n",
       "      <td>2697</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>LTR/Gypsy</td>\n",
       "      <td>SO:0002265</td>\n",
       "      <td>0.836</td>\n",
       "      <td>homology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ptg000002l_3</td>\n",
       "      <td>ptg000002l</td>\n",
       "      <td>EDTA</td>\n",
       "      <td>Gypsy_LTR_retrotransposon</td>\n",
       "      <td>20481</td>\n",
       "      <td>24091</td>\n",
       "      <td>9089</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>LTR/Gypsy</td>\n",
       "      <td>SO:0002265</td>\n",
       "      <td>0.883</td>\n",
       "      <td>homology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ptg000002l_4</td>\n",
       "      <td>ptg000002l</td>\n",
       "      <td>EDTA</td>\n",
       "      <td>Gypsy_LTR_retrotransposon</td>\n",
       "      <td>24092</td>\n",
       "      <td>24877</td>\n",
       "      <td>5031</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>LTR/Gypsy</td>\n",
       "      <td>SO:0002265</td>\n",
       "      <td>0.872</td>\n",
       "      <td>homology</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          te_id       seqid source          sequence_ontology  start    end  \\\n",
       "0  ptg000002l_0  ptg000002l   EDTA  Gypsy_LTR_retrotransposon  15839  17218   \n",
       "1  ptg000002l_1  ptg000002l   EDTA  Gypsy_LTR_retrotransposon  17520  19985   \n",
       "2  ptg000002l_2  ptg000002l   EDTA  Gypsy_LTR_retrotransposon  19987  20451   \n",
       "3  ptg000002l_3  ptg000002l   EDTA  Gypsy_LTR_retrotransposon  20481  24091   \n",
       "4  ptg000002l_4  ptg000002l   EDTA  Gypsy_LTR_retrotransposon  24092  24877   \n",
       "\n",
       "   score strand phase classification sequence_ontology_v2 identity    method  \n",
       "0   4477      +     .      LTR/Gypsy           SO:0002265    0.753  homology  \n",
       "1  12445      +     .      LTR/Gypsy           SO:0002265    0.835  homology  \n",
       "2   2697      +     .      LTR/Gypsy           SO:0002265    0.836  homology  \n",
       "3   9089      +     .      LTR/Gypsy           SO:0002265    0.883  homology  \n",
       "4   5031      +     .      LTR/Gypsy           SO:0002265    0.872  homology  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Reading TEs file and preparing TE collection\n",
    "te_anno_headers = ['seqid','source','sequence_ontology', 'start', \n",
    "                   'end', 'score', 'strand', 'phase', 'attributes']\n",
    "te_anno = pd.read_csv('./ptg000002l.fasta.TEanno.gff3.gz', sep='\\t', skiprows=6, header=None)\n",
    "te_anno.columns = te_anno_headers\n",
    "te_ids = []\n",
    "for i in range(len(te_anno)):\n",
    "    attributes_info = te_anno.iloc[i][\"attributes\"].split(';')\n",
    "    te_anno.at[i,'classification'] = attributes_info[2].split('=')[1]\n",
    "    te_anno.at[i,'sequence_ontology_v2'] = attributes_info[3].split('=')[1]\n",
    "    te_anno.at[i,'identity'] = attributes_info[4].split('=')[1]\n",
    "    te_anno.at[i,'method'] = attributes_info[5].split('=')[1]\n",
    "    te_ids.append(te_anno.iloc[i][\"seqid\"]+'_'+str(i))\n",
    "del te_anno['attributes']\n",
    "te_anno.insert(0, 'te_id', te_ids)\n",
    "te_anno.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8cd18efd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reading chromosome\n",
    "chromosome = ''\n",
    "for record in SeqIO.parse(\"./ptg000002l.fasta\", \"fasta\"):\n",
    "    chromosome = record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ef69e49d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Adding TEs sequences to df\n",
    "for i in range(len(te_anno)):\n",
    "    start = te_anno.iloc[i]['start']\n",
    "    end = te_anno.iloc[i]['end']+1\n",
    "    te_anno.at[i, 'sequence'] = str(chromosome.seq[start:end])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ece171e",
   "metadata": {},
   "source": [
    "### Saving TEs records in mongo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "03a037ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pymongo.results.InsertManyResult at 0x7f113ebe63c0>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ##Saving TEs records\n",
    "collection = db['tes']\n",
    "tes_dict = te_anno.to_dict(\"records\")\n",
    "collection.insert_many(tes_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee2ec6ec",
   "metadata": {},
   "source": [
    "# Genes and proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "734f98ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "curl: (35) OpenSSL SSL_connect: Connection reset by peer in connection to eutils.ncbi.nlm.nih.gov:443 \n",
      "\u001b[m\u001b[?7h\u001b[4l\u001b>\u001b7\u001b[r\u001b[?1;3;4;6l\u001b8\u001b[31m\u001b[1m\u001b[7m ERROR: \u001b[m\u001b[?7h\u001b[4l\u001b>\u001b7\u001b[r\u001b[?1;3;4;6l\u001b8\u001b[31m\u001b[1m curl command failed ( Wed Oct 19 02:33:04 PM -05 2022 ) with: 35\u001b[m\u001b[?7h\u001b[4l\u001b>\u001b7\u001b[r\u001b[?1;3;4;6l\u001b8\n",
      "\u001b[34mhttps://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi -d retmax=0&usehistory=y&db=protein&term=XP_002512792.2&tool=edirect&edirect=16.2&edirect_os=Linux&email=breaze%40breaze\u001b[m\u001b[?7h\u001b[4l\u001b>\u001b7\u001b[r\u001b[?1;3;4;6l\u001b8\n",
      "\u001b[m\u001b[31m\u001b[1m\u001b[7m WARNING: \u001b[m\u001b[31m\u001b[1m FAILURE ( Wed Oct 19 02:33:04 PM -05 2022 )\u001b[m\n",
      "\u001b[34mnquire -url https://eutils.ncbi.nlm.nih.gov/entrez/eutils/ esearch.fcgi -retmax 0 -usehistory y -db protein -term XP_002512792.2 -tool edirect -edirect 16.2 -edirect_os Linux -email breaze@breaze\n",
      "\u001b[m\u001b[31mEMPTY RESULT\u001b[m\n",
      "\u001b[34mSECOND ATTEMPT\u001b[m\n",
      "curl: (28) SSL connection timeout\n",
      "\u001b[m\u001b[?7h\u001b[4l\u001b>\u001b7\u001b[r\u001b[?1;3;4;6l\u001b8\u001b[31m\u001b[1m\u001b[7m ERROR: \u001b[m\u001b[?7h\u001b[4l\u001b>\u001b7\u001b[r\u001b[?1;3;4;6l\u001b8\u001b[31m\u001b[1m curl command failed ( Wed Oct 19 02:55:55 PM -05 2022 ) with: 28\u001b[m\u001b[?7h\u001b[4l\u001b>\u001b7\u001b[r\u001b[?1;3;4;6l\u001b8\n",
      "\u001b[34mhttps://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi -d retmax=0&usehistory=y&db=protein&term=PKA65666.1&tool=edirect&edirect=16.2&edirect_os=Linux&email=breaze%40breaze\u001b[m\u001b[?7h\u001b[4l\u001b>\u001b7\u001b[r\u001b[?1;3;4;6l\u001b8\n",
      "\u001b[m\u001b[31m\u001b[1m\u001b[7m WARNING: \u001b[m\u001b[31m\u001b[1m FAILURE ( Wed Oct 19 02:55:55 PM -05 2022 )\u001b[m\n",
      "\u001b[34mnquire -url https://eutils.ncbi.nlm.nih.gov/entrez/eutils/ esearch.fcgi -retmax 0 -usehistory y -db protein -term PKA65666.1 -tool edirect -edirect 16.2 -edirect_os Linux -email breaze@breaze\n",
      "\u001b[m\u001b[31mEMPTY RESULT\u001b[m\n",
      "\u001b[34mSECOND ATTEMPT\u001b[m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>blast_prot_pred_id</th>\n",
       "      <th>query_accession</th>\n",
       "      <th>target_accession</th>\n",
       "      <th>sequence_identity</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gap_openings</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>target_start</th>\n",
       "      <th>target_end</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>gb_info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4.t1_0</td>\n",
       "      <td>g4.t1</td>\n",
       "      <td>XP_031739819.1</td>\n",
       "      <td>84.9</td>\n",
       "      <td>225</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>224</td>\n",
       "      <td>1</td>\n",
       "      <td>225</td>\n",
       "      <td>8.400000e-104</td>\n",
       "      <td>388.3</td>\n",
       "      <td>LOCUS       XP_030957800    1121 aa   PROTEIN ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4.t1_1</td>\n",
       "      <td>g4.t1</td>\n",
       "      <td>KAH7528545.1</td>\n",
       "      <td>81.6</td>\n",
       "      <td>147</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>147</td>\n",
       "      <td>9</td>\n",
       "      <td>155</td>\n",
       "      <td>2.700000e-62</td>\n",
       "      <td>250.4</td>\n",
       "      <td>LOCUS       XP_030957800    1121 aa   PROTEIN ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g6.t1_2</td>\n",
       "      <td>g6.t1</td>\n",
       "      <td>PNS97074.1</td>\n",
       "      <td>80.2</td>\n",
       "      <td>329</td>\n",
       "      <td>63</td>\n",
       "      <td>2</td>\n",
       "      <td>109</td>\n",
       "      <td>437</td>\n",
       "      <td>1</td>\n",
       "      <td>327</td>\n",
       "      <td>7.400000e-157</td>\n",
       "      <td>564.7</td>\n",
       "      <td>LOCUS       XP_030957800    1121 aa   PROTEIN ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g7.t1_3</td>\n",
       "      <td>g7.t1</td>\n",
       "      <td>MBA0694754.1</td>\n",
       "      <td>80.4</td>\n",
       "      <td>465</td>\n",
       "      <td>68</td>\n",
       "      <td>2</td>\n",
       "      <td>1105</td>\n",
       "      <td>1567</td>\n",
       "      <td>1</td>\n",
       "      <td>444</td>\n",
       "      <td>1.200000e-220</td>\n",
       "      <td>778.9</td>\n",
       "      <td>LOCUS       XP_030957800    1121 aa   PROTEIN ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g8.t1_4</td>\n",
       "      <td>g8.t1</td>\n",
       "      <td>KAH7572690.1</td>\n",
       "      <td>80.0</td>\n",
       "      <td>105</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>105</td>\n",
       "      <td>1</td>\n",
       "      <td>105</td>\n",
       "      <td>3.200000e-40</td>\n",
       "      <td>175.3</td>\n",
       "      <td>LOCUS       XP_030957800    1121 aa   PROTEIN ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  blast_prot_pred_id query_accession target_accession  sequence_identity  \\\n",
       "0            g4.t1_0           g4.t1   XP_031739819.1               84.9   \n",
       "1            g4.t1_1           g4.t1     KAH7528545.1               81.6   \n",
       "2            g6.t1_2           g6.t1       PNS97074.1               80.2   \n",
       "3            g7.t1_3           g7.t1     MBA0694754.1               80.4   \n",
       "4            g8.t1_4           g8.t1     KAH7572690.1               80.0   \n",
       "\n",
       "   length  mismatches  gap_openings  query_start  query_end  target_start  \\\n",
       "0     225          33             1            1        224             1   \n",
       "1     147          27             0            1        147             9   \n",
       "2     329          63             2          109        437             1   \n",
       "3     465          68             2         1105       1567             1   \n",
       "4     105          21             0            1        105             1   \n",
       "\n",
       "   target_end        e_value  bit_score  \\\n",
       "0         225  8.400000e-104      388.3   \n",
       "1         155   2.700000e-62      250.4   \n",
       "2         327  7.400000e-157      564.7   \n",
       "3         444  1.200000e-220      778.9   \n",
       "4         105   3.200000e-40      175.3   \n",
       "\n",
       "                                             gb_info  \n",
       "0  LOCUS       XP_030957800    1121 aa   PROTEIN ...  \n",
       "1  LOCUS       XP_030957800    1121 aa   PROTEIN ...  \n",
       "2  LOCUS       XP_030957800    1121 aa   PROTEIN ...  \n",
       "3  LOCUS       XP_030957800    1121 aa   PROTEIN ...  \n",
       "4  LOCUS       XP_030957800    1121 aa   PROTEIN ...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Reading blast with predicted proteins\n",
    "blast_prot = pd.read_csv('./Blast_NR.abinitio_prot_Dm.tab', sep='\\t', header=None)\n",
    "blast_prot_headers = ['query_accession', 'target_accession', 'sequence_identity', 'length', 'mismatches', \n",
    "                      'gap_openings', 'query_start', 'query_end', 'target_start', 'target_end', \n",
    "                     'e_value', 'bit_score']\n",
    "blast_prot.columns = blast_prot_headers\n",
    "blast_predicted_proteins_ids = []\n",
    "for i in range(len(blast_prot)):\n",
    "    query = blast_prot.iloc[i][\"target_accession\"]\n",
    "    !esearch -db protein -query {query} | efetch -format gb > result.gb\n",
    "    with open(\"result.gb\") as handle:\n",
    "        try:\n",
    "            record = GenBank.read(handle)\n",
    "            blast_prot[\"gb_info\"] = str(record)\n",
    "        except:\n",
    "            blast_prot[\"gb_info\"] = \"\"\n",
    "    blast_predicted_proteins_ids.append(blast_prot.iloc[i][\"query_accession\"]+'_'+str(i))\n",
    "blast_prot.insert(0, 'blast_prot_pred_id', blast_predicted_proteins_ids)\n",
    "blast_prot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5305e8ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pymongo.results.InsertManyResult at 0x7f11ab026980>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ##Saving records of blast with predicted proteins \n",
    "collection = db['blast_predicted_proteins']\n",
    "blast_pred_prot_dict = blast_prot.to_dict(\"records\")\n",
    "collection.insert_many(blast_pred_prot_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4bbe85d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prot_domain_id</th>\n",
       "      <th>seqid</th>\n",
       "      <th>alignment_start</th>\n",
       "      <th>alignment_end</th>\n",
       "      <th>envelope_start</th>\n",
       "      <th>envelope_end</th>\n",
       "      <th>hmm_acc</th>\n",
       "      <th>hmm_name</th>\n",
       "      <th>type</th>\n",
       "      <th>hmm_start</th>\n",
       "      <th>hmm_end</th>\n",
       "      <th>hmm_length</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>e_value</th>\n",
       "      <th>significance</th>\n",
       "      <th>clan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4.t1_0</td>\n",
       "      <td>g4.t1</td>\n",
       "      <td>173</td>\n",
       "      <td>313</td>\n",
       "      <td>154</td>\n",
       "      <td>313</td>\n",
       "      <td>PF13359.9</td>\n",
       "      <td>DDE_Tnp_4</td>\n",
       "      <td>Domain</td>\n",
       "      <td>24</td>\n",
       "      <td>158</td>\n",
       "      <td>158</td>\n",
       "      <td>51.9</td>\n",
       "      <td>7.200000e-14</td>\n",
       "      <td>1</td>\n",
       "      <td>CL0219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g5.t1_1</td>\n",
       "      <td>g5.t1</td>\n",
       "      <td>52</td>\n",
       "      <td>119</td>\n",
       "      <td>52</td>\n",
       "      <td>144</td>\n",
       "      <td>PF00300.25</td>\n",
       "      <td>His_Phos_1</td>\n",
       "      <td>Domain</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>194</td>\n",
       "      <td>29.8</td>\n",
       "      <td>4.700000e-07</td>\n",
       "      <td>1</td>\n",
       "      <td>CL0071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g6.t1_2</td>\n",
       "      <td>g6.t1</td>\n",
       "      <td>64</td>\n",
       "      <td>260</td>\n",
       "      <td>61</td>\n",
       "      <td>260</td>\n",
       "      <td>PF05634.14</td>\n",
       "      <td>APO_RNA-bind</td>\n",
       "      <td>Family</td>\n",
       "      <td>4</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>347.5</td>\n",
       "      <td>1.800000e-104</td>\n",
       "      <td>1</td>\n",
       "      <td>No_clan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g6.t1_3</td>\n",
       "      <td>g6.t1</td>\n",
       "      <td>300</td>\n",
       "      <td>415</td>\n",
       "      <td>291</td>\n",
       "      <td>427</td>\n",
       "      <td>PF05634.14</td>\n",
       "      <td>APO_RNA-bind</td>\n",
       "      <td>Family</td>\n",
       "      <td>66</td>\n",
       "      <td>181</td>\n",
       "      <td>200</td>\n",
       "      <td>82.2</td>\n",
       "      <td>3.300000e-23</td>\n",
       "      <td>1</td>\n",
       "      <td>No_clan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g7.t1_4</td>\n",
       "      <td>g7.t1</td>\n",
       "      <td>144</td>\n",
       "      <td>190</td>\n",
       "      <td>144</td>\n",
       "      <td>192</td>\n",
       "      <td>PF13041.9</td>\n",
       "      <td>PPR_2</td>\n",
       "      <td>Repeat</td>\n",
       "      <td>1</td>\n",
       "      <td>47</td>\n",
       "      <td>50</td>\n",
       "      <td>31.6</td>\n",
       "      <td>1.500000e-07</td>\n",
       "      <td>1</td>\n",
       "      <td>CL0020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  prot_domain_id  seqid  alignment_start  alignment_end  envelope_start  \\\n",
       "0        g4.t1_0  g4.t1              173            313             154   \n",
       "1        g5.t1_1  g5.t1               52            119              52   \n",
       "2        g6.t1_2  g6.t1               64            260              61   \n",
       "3        g6.t1_3  g6.t1              300            415             291   \n",
       "4        g7.t1_4  g7.t1              144            190             144   \n",
       "\n",
       "   envelope_end     hmm_acc      hmm_name    type  hmm_start  hmm_end  \\\n",
       "0           313   PF13359.9     DDE_Tnp_4  Domain         24      158   \n",
       "1           144  PF00300.25    His_Phos_1  Domain          1       68   \n",
       "2           260  PF05634.14  APO_RNA-bind  Family          4      200   \n",
       "3           427  PF05634.14  APO_RNA-bind  Family         66      181   \n",
       "4           192   PF13041.9         PPR_2  Repeat          1       47   \n",
       "\n",
       "   hmm_length  bit_score        e_value  significance     clan  \n",
       "0         158       51.9   7.200000e-14             1   CL0219  \n",
       "1         194       29.8   4.700000e-07             1   CL0071  \n",
       "2         200      347.5  1.800000e-104             1  No_clan  \n",
       "3         200       82.2   3.300000e-23             1  No_clan  \n",
       "4          50       31.6   1.500000e-07             1   CL0020  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Reading pfam file with protein domains\n",
    "domains = pd.read_csv('./augustus.abinitio.pfam_results.txt', delim_whitespace=True, skiprows=28, header=None)\n",
    "domains_headers = ['seqid','alignment_start','alignment_end', 'envelope_start', \n",
    "                   'envelope_end', 'hmm_acc', 'hmm_name', 'type', 'hmm_start', 'hmm_end', 'hmm_length',\n",
    "                  'bit_score', 'e_value', 'significance', 'clan']\n",
    "domains.columns = domains_headers\n",
    "prot_domain_ids = []\n",
    "for i in range(len(domains)):\n",
    "    prot_domain_ids.append(domains.iloc[i][\"seqid\"]+'_'+str(i))\n",
    "domains.insert(0, 'prot_domain_id', prot_domain_ids)\n",
    "domains.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "edff115d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pymongo.results.InsertManyResult at 0x7f11aadd4dc0>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##Saving protein domains\n",
    "collection = db['protein_domains']\n",
    "domains_dict = domains.to_dict(\"records\")\n",
    "collection.insert_many(domains_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0fb6c234",
   "metadata": {},
   "outputs": [],
   "source": [
    "def searchBlastPredictedProteins(query):\n",
    "    collection = db['blast_predicted_proteins']\n",
    "    blast_pred_prot_query = {\"query_accession\": query}\n",
    "    mydoc = collection.find(blast_pred_prot_query)\n",
    "    blast_pred_prot_ref = []\n",
    "    for bl in mydoc:\n",
    "        blast_pred_prot_ref.append(bl[\"_id\"])\n",
    "    return blast_pred_prot_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "cb35ef4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def searchProteinDomain(query):\n",
    "    collection = db['protein_domains']\n",
    "    protein_domain_query = {\"seqid\": query}\n",
    "    mydoc = collection.find(protein_domain_query)\n",
    "    protein_domain_ref = []\n",
    "    for prt in mydoc:\n",
    "        protein_domain_ref.append(prt[\"_id\"])\n",
    "    return protein_domain_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a71bc5ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gen_id</th>\n",
       "      <th>seq</th>\n",
       "      <th>blast_predicted_prot</th>\n",
       "      <th>protein_domains</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g1.t1</td>\n",
       "      <td>MDQATELEREDNALLKQFLDLKPIKYKGIGNASLAEDWLAEMNKIF...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g2.t1</td>\n",
       "      <td>MKVHVPDSVGLTTVLHSVWKSEVHIPDSAGPTTALHSEWHDRIRLG...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g3.t1</td>\n",
       "      <td>MRLHVPDSAGPTTALQSEKVHVPDSAGLTTALHSEWHVPDSAGLTT...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g4.t1</td>\n",
       "      <td>MESSDDEKDLLCGKSLVKDLAYNLVPTGIKFVDEVLNGPNERCLEN...</td>\n",
       "      <td>[635059954621c31f518fbb39, 635059954621c31f518...</td>\n",
       "      <td>[63505a084621c31f518fc8f2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g5.t1</td>\n",
       "      <td>MHPAACKVQPLPLFCPKSIRKSKSTKSLRLVTETPTAEDQASPSPI...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[63505a084621c31f518fc8f3]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gen_id                                                seq  \\\n",
       "0  g1.t1  MDQATELEREDNALLKQFLDLKPIKYKGIGNASLAEDWLAEMNKIF...   \n",
       "1  g2.t1  MKVHVPDSVGLTTVLHSVWKSEVHIPDSAGPTTALHSEWHDRIRLG...   \n",
       "2  g3.t1  MRLHVPDSAGPTTALQSEKVHVPDSAGLTTALHSEWHVPDSAGLTT...   \n",
       "3  g4.t1  MESSDDEKDLLCGKSLVKDLAYNLVPTGIKFVDEVLNGPNERCLEN...   \n",
       "4  g5.t1  MHPAACKVQPLPLFCPKSIRKSKSTKSLRLVTETPTAEDQASPSPI...   \n",
       "\n",
       "                                blast_predicted_prot  \\\n",
       "0                                                 []   \n",
       "1                                                 []   \n",
       "2                                                 []   \n",
       "3  [635059954621c31f518fbb39, 635059954621c31f518...   \n",
       "4                                                 []   \n",
       "\n",
       "              protein_domains  \n",
       "0                          []  \n",
       "1                          []  \n",
       "2                          []  \n",
       "3  [63505a084621c31f518fc8f2]  \n",
       "4  [63505a084621c31f518fc8f3]  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Reading and processing genes collection\n",
    "genes = pd.DataFrame()\n",
    "genes_ids = []\n",
    "genes_seq = []\n",
    "for gen in SeqIO.parse(\"./augustus.abinitio.aa\", \"fasta\"):\n",
    "    genes_ids.append(gen.id)\n",
    "    genes_seq.append(str(gen.seq))\n",
    "genes[\"gen_id\"] = genes_ids\n",
    "genes[\"seq\"] = genes_seq\n",
    "temp_blast_refs = []\n",
    "temp_domain_refs = []\n",
    "for i in range(len(genes)):\n",
    "    temp_blast_refs.append(searchBlastPredictedProteins(genes.iloc[i][\"gen_id\"]))\n",
    "    temp_domain_refs.append(searchProteinDomain(genes.iloc[i][\"gen_id\"]))\n",
    "genes[\"blast_predicted_prot\"] = temp_blast_refs\n",
    "genes[\"protein_domains\"] = temp_domain_refs\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "05a95f66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pymongo.results.InsertManyResult at 0x7f11aaa7c200>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##Saving records predicted genes\n",
    "collection = db['predicted_genes']\n",
    "predicted_genes_dict = genes.to_dict(\"records\")\n",
    "collection.insert_many(predicted_genes_dict)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
